<h3 style="text-align:center">Kel 3 - Implementasi Library Google Cloud Vision untuk ekstraksi karakter pada plat nomor kendaraann pada seluruh gambar baik pada gambar yang belum di filtering dengan metode sauvola maupun yang sudah difiltering.</h3>



Tahap pertama ialah mengimpor library dan mencantumkan api key dalam bentuk json yang sebelumnya telah dibuat di akun google cloud platform untuk API google vision.

In [1]:

import os
import math
from collections import Counter
from google.cloud import vision
import re

os.environ['GOOGLE_APPLICATION_CREDENTIALS'] ='dicoding-architect-gcp-be711cac9be0.json'
WORD = re.compile(r"\w+")

Tahap kedua ialah mendefinisikan fungsi detect_text dimana pada fungsi ini, path dari sebuah gambar yang kita gunakan akan dikenali dan akan dibaca dalam bentuk byte string. metode document_text_detection Google Vision API digunaka untuk mendeteksi teks dalam dokumen atau gambar dengan teks yang terstruktur.Lalu, text_annotations Mengambil hasil OCR dari respon API yang kemudian hasilnya dimasukkan ke dalam list ocr_text.

In [2]:
def detect_text(path):
    """Detects text in the file."""

    client = vision.ImageAnnotatorClient()

    with open(path, "rb") as image_file:
        content = image_file.read()

    image = vision.Image(content=content)

   
    response = client.document_text_detection(image=image)
    texts = response.text_annotations
    ocr_text = []

    for text in texts:
        ocr_text.append(f"\r\n{text.description}")

    if response.error.message:
        raise Exception(
            "{}\nFor more info on error messages, check: "
            "https://cloud.google.com/apis/design/errors".format(response.error.message)
        )
    return ocr_text

Mendefinisikan path folder gambar input untuk gambar yang belum difiltering dengan metode sauvola. Pada tahap ini juga akan didefinisikan list yang berfungsi untuk menyimpan output.

In [23]:
# Path folder gambar
input_folder = "./input_100"

# Dictionary untuk menyimpan hasil deteksi teks
detected_texts_non_sauvola = {}

Mendefinisikan path folder gambar input untuk gambar yang sudah difiltering dengan metode sauvola. Pada tahap ini juga akan didefinisikan list yang berfungsi untuk menyimpan output.

In [53]:
# Path folder gambar
input_folder = "./output_100"

# Dictionary untuk menyimpan hasil deteksi teks
detected_texts_sauvola = {}

Dikarenakan gamabr yang digunakan cukup banyak dan jika dilakukan manual akan cukup lama maka akan dilakukan perulangan sebanyak jumlah gambar yang ada di folder. Setiap gambar dalam perulangan tersebut akan dideteksi textnya dan hasilnya akan dimasukkan ke dalam dictionary dengan key berupa nama gambarnya sedangkan valuenya ialah hasil dari ocrnya.

In [54]:
# Proses setiap file gambar dalam folder
for file_name in os.listdir(input_folder):
    if file_name.endswith(('.jpg', '.jpeg', '.png')):  # Filter hanya file gambar
        image_path = os.path.join(input_folder, file_name)
        
        try:
            # Deteksi teks pada gambar
            text = detect_text(image_path)
            detected_texts_sauvola[file_name] = text

            # Mencetak hasil deteksi teks untuk setiap gambar
            print(f"\nDetected text for image: {file_name}")
            for line in text:
                print(line)  # Mencetak setiap baris teks yang terdeteksi

        except Exception as e:
            print(f"Error processing {file_name}: {e}")
            detected_texts_sauvola[file_name] = None  # Jika error, simpan None


Detected text for image: gambar_filtering_1.jpg

B 5430-THW
051-26

B

5430

-

THW

051-26

Detected text for image: gambar_filtering_10.jpg

عام
C
со
མ
5
on
N
C

عام

C

со

མ

5

on

N

C

Detected text for image: gambar_filtering_100.jpg

B.6408 KWV
06W25

B.6408

KWV

06W25

Detected text for image: gambar_filtering_101.jpg

wi
5470 FEE
07-26

wi

5470

FEE

07-26

Detected text for image: gambar_filtering_102.jpg

TOKY

TOKY

Detected text for image: gambar_filtering_103.jpg

B 4355 TYI
10.28

B

4355

TYI

10.28

Detected text for image: gambar_filtering_104.jpg

B 1064 HUB
11-24

B

1064

HUB

11-24

Detected text for image: gambar_filtering_11.jpg

B 5961 FPH
11-28
CAS POURI

B

5961

FPH

11-28

CAS

POURI

Detected text for image: gambar_filtering_12.jpg

B 6685 T80
0128
AXORLANTAS POLRI

B

6685

T80

0128

AXORLANTAS

POLRI

Detected text for image: gambar_filtering_13.jpg

B 5351 KAY
=
"
12-27
☑
"
"
-
☐

B

5351

KAY

=

"

12-27

☑

"

"

-

☐

Detected text for image: 

Kode dibawah ialah contoh isi dari dictionary untuk hasil ocr pada seluruh gambar yang belum difiltering oleh metode sauvola.

In [25]:

print("\nFinal Detected Texts Dictionary:")
print(detected_texts_non_sauvola)


Final Detected Texts Dictionary:
{'gambar_1.jpg': ['\r\nB 5430-THW\n05-26', '\r\nB', '\r\n5430', '\r\n-', '\r\nTHW', '\r\n05-26'], 'gambar_10.jpg': ['\r\nN\n2\n6\nORLANTAS POLRI\nב\nB\nC', '\r\nN', '\r\n2', '\r\n6', '\r\nORLANTAS', '\r\nPOLRI', '\r\nב', '\r\nB', '\r\nC'], 'gambar_100.jpg': ['\r\nB.6408 KWV\nEXC\n06-25', '\r\nB.6408', '\r\nKWV', '\r\nEXC', '\r\n06-25'], 'gambar_101.jpg': ['\r\nB 5470 FEF\nHONDA\n07-26', '\r\nB', '\r\n5470', '\r\nFEF', '\r\nHONDA', '\r\n07-26'], 'gambar_102.jpg': ['\r\nB 3715 KOY\n07-24', '\r\nB', '\r\n3715', '\r\nKOY', '\r\n07-24'], 'gambar_103.jpg': ['\r\nB 4355 TYI\nHON\n10-28', '\r\nB', '\r\n4355', '\r\nTYI', '\r\nHON', '\r\n10-28'], 'gambar_104.jpg': ['\r\nB 1064 HUB\n11-29', '\r\nB', '\r\n1064', '\r\nHUB', '\r\n11-29'], 'gambar_11.jpg': ['\r\nB 5961 FPH\n11.28', '\r\nB', '\r\n5961', '\r\nFPH', '\r\n11.28'], 'gambar_12.jpg': ['\r\nB 6685 TOO\n01-28\nRORLANTAS POLRI', '\r\nB', '\r\n6685', '\r\nTOO', '\r\n01-28', '\r\nRORLANTAS', '\r\nPOLRI'], 'gamba

Untuk mempermudah melihat data maka lebih baik jika dictionary tersebut diubah dalam bentuk dataframe python.<br>
<br>
<br>
penjelasan hasil ocr, jika hasil ocr berupa '\r\nB 5430-THW\n05-26', maka simbol '\r' dan '\n' merupakan spasi, yang berarti hasil ocrnya ialah <b>B 5430-THW 05-26

In [39]:
import pandas as pd

# Jika data Anda berbentuk tabular atau dapat dijadikan DataFrame
df = pd.DataFrame.from_dict(detected_texts_non_sauvola, orient='index')
print(df)


                                                                                                     0   \
gambar_1.jpg                                                                      \r\nB 5430-THW\n05-26   
gambar_10.jpg                                                      \r\nN\n2\n6\nORLANTAS POLRI\nב\nB\nC   
gambar_100.jpg                                                               \r\nB.6408 KWV\nEXC\n06-25   
gambar_101.jpg                                                             \r\nB 5470 FEF\nHONDA\n07-26   
gambar_102.jpg                                                                    \r\nB 3715 KOY\n07-24   
gambar_103.jpg                                                               \r\nB 4355 TYI\nHON\n10-28   
gambar_104.jpg                                                                    \r\nB 1064 HUB\n11-29   
gambar_11.jpg                                                                     \r\nB 5961 FPH\n11.28   
gambar_12.jpg                        

In [41]:
from IPython.display import display

display(df)  # Menampilkan DataFrame dengan format lebih baik di Jupyter Notebook

0   \
gambar_1.jpg                                                                      \r\nB 5430-THW\n05-26   
gambar_10.jpg                                                      \r\nN\n2\n6\nORLANTAS POLRI\nב\nB\nC   
gambar_100.jpg                                                               \r\nB.6408 KWV\nEXC\n06-25   
gambar_101.jpg                                                             \r\nB 5470 FEF\nHONDA\n07-26   
gambar_102.jpg                                                                    \r\nB 3715 KOY\n07-24   
gambar_103.jpg                                                               \r\nB 4355 TYI\nHON\n10-28   
gambar_104.jpg                                                                    \r\nB 1064 HUB\n11-29   
gambar_11.jpg                                                                     \r\nB 5961 FPH\n11.28   
gambar_12.jpg                                                    \r\nB 6685 TOO\n01-28\nRORLANTAS POLRI   
gambar_13.jpg                                                              \r\nB 5351 KAY\nHONDA\n12.27   
gambar_14.jpg                                                                     \r\nB 4399 KIX\n06.28   
gambar_15.jpg                                   \r\nB 4215 KZR\nMA\nMOTOR\nSHOP\n08-27\nKORLANTAS POLRI   
gambar_16.jpg                                                                     \r\nB 5526 TNK\n03.28   
gambar_17.jpg                                              \r\nA\n2\nORLANTAS POLRI\nKORLANTAS POLRI\nH   
gambar_18.jpg                                                              \r\nB 5560 KJR\nHONDA\n10.29   
gambar_19.jpg                                                       \r\nKOLANTAS POLRI\nORLANTAS POL\nS   
gambar_2.jpg     \r\nB 5383 FUF\nYAMAHA\nXTW\nLIVE IN THE HIGHER STAGE\nYAMAHA\n08.29\nSEMAKIN DI DEPAN   
gambar_20.jpg                                                              \r\nB 4131 KMN\nHONDA\n04.29   
gambar_21.jpg                                                                \r\nB 5171 FOV\nMAX\n10.28   
gambar_22.jpg                                                               \r\nB 4642 TLK\nHONDA 06.26   
gambar_23.jpg     \r\nB 5859 FLE\nMAX\nYAMAHA\nLIVE IN THE HIGHER STAGE\nYAMAHA\n02.28\nKORLANTAS POLRI   
gambar_24.jpg                                                              \r\nB 5395 FFE\nHONDA\n10-26   
gambar_25.jpg                                 \r\nB 5199 FSP\nVAMA\nMX\nLIVE IN THE HIGHER STAGE\n05-29   
gambar_26.jpg                                              \r\nB 5946 FCB\nREMAS SUS\n02-26\nVAMAHA\n00   
gambar_27.jpg                                                              \r\nB 5039 TLX\nHONDA\n10.27   
gambar_28.jpg                                                              \r\nB 4318 FFB\nHONDA\n06.26   
gambar_29.jpg                                                     \r\nB 4542 TPX\nYOSHIMURA 3347\n06.28   
gambar_3.jpg                                                              \r\nQ\nU\n10.28\nKOREAN POLRI   
gambar_30.jpg                                                    \r\nB 3416 TPL\n09.27\nKORLANTAS POLRT   
gambar_31.jpg                                                                     \r\nB 4739 FWD\n08-29   
gambar_32.jpg                                                           \r\nB 5100 FLW\nHONDA\n12.28\nS   
gambar_33.jpg                                                             \r\nB 4238 KIW\nHONDA\n106.28   
gambar_34.jpg                                                              \r\nB 4256 KEH\nMinone 09.27   
gambar_35.jpg                                                              \r\nB 4580 KXX\nHONDA\n01.28   
gambar_36.jpg                                                           \r\nB 4577 KGZ\nHONDA\n12-27\nB   
gambar_37.jpg                                                             \r\nB 5155 TKG\nHONDAP\n03.27   
gambar_38.jpg                                                           \r\nB 4957 KWP\nBIG WING\n03.27   
gambar_39.jpg                      

Hal yang sama dilakukan pada seluruh gambar yang telah difiltering dengan metode sauvola.

In [55]:
import pandas as pd

# Jika data Anda berbentuk tabular atau dapat dijadikan DataFrame
df_sauvola = pd.DataFrame.from_dict(detected_texts_sauvola, orient='index')
print(df_sauvola)


                                                                                                                     0   \
gambar_filtering_1.jpg                                                                           \r\nB 5430-THW\n051-26   
gambar_filtering_10.jpg                                                                  \r\nعام\nC\nсо\nམ\n5\non\nN\nC   
gambar_filtering_100.jpg                                                                          \r\nB.6408 KWV\n06W25   
gambar_filtering_101.jpg                                                                        \r\nwi\n5470 FEE\n07-26   
gambar_filtering_102.jpg                                                                                       \r\nTOKY   
gambar_filtering_103.jpg                                                                          \r\nB 4355 TYI\n10.28   
gambar_filtering_104.jpg                                                                          \r\nB 1064 HUB\n11-24   
gambar_filtering

In [56]:
from IPython.display import display

display(df_sauvola)  # Menampilkan DataFrame dengan format lebih baik di Jupyter Notebook

0   \
gambar_filtering_1.jpg                                                                           \r\nB 5430-THW\n051-26   
gambar_filtering_10.jpg                                                                  \r\nعام\nC\nсо\nམ\n5\non\nN\nC   
gambar_filtering_100.jpg                                                                          \r\nB.6408 KWV\n06W25   
gambar_filtering_101.jpg                                                                        \r\nwi\n5470 FEE\n07-26   
gambar_filtering_102.jpg                                                                                       \r\nTOKY   
gambar_filtering_103.jpg                                                                          \r\nB 4355 TYI\n10.28   
gambar_filtering_104.jpg                                                                          \r\nB 1064 HUB\n11-24   
gambar_filtering_11.jpg                                                                \r\nB 5961 FPH\n11-28\nCAS POURI   
gambar_filtering_12.jpg                                                          \r\nB 6685 T80\n0128\nAXORLANTAS POLRI   
gambar_filtering_13.jpg                                                      \r\nB 5351 KAY\n=\n"\n12-27\n☑\n"\n"\n-\n☐   
gambar_filtering_14.jpg                                                                           \r\nB 4399 KIX\n06.28   
gambar_filtering_15.jpg                          \r\n亠\nB (215 KZR\nMA\nMOTOR\nSHOPY\n"\nr\n08-27\nPOLR)\nJUNYT SLOW\n"   
gambar_filtering_16.jpg                                                                          \r\nB 5526 TNK\n03° 28   
gambar_filtering_17.jpg                                                             \r\nKORLANT\nKORI ANTAS POLRI\n1\nR   
gambar_filtering_18.jpg                                                                   \r\nபி\nB 5560 KJR\nHONDA\n10   
gambar_filtering_19.jpg   \r\n3.t\n་་\n14\n2.17\nな\n*⌘、 “!{S!!་\nC\n☐\nO\n& th\nANTAST\n31\nי\n"\nKOW-ANTAS POLRI\n_\nC   
gambar_filtering_2.jpg                                           \r\nB 5383 FUF\n08.29\nYAMAHA\nYAMAHA\nSEMAKIN DIDEPAN   
gambar_filtering_20.jpg                                                                               \r\nKORVANTAS POL   
gambar_filtering_21.jpg                                                                  \r\nB 5171 FOY\n10.28\n18106 S   
gambar_filtering_22.jpg                                                            \r\nORL\n1\n1\n"\nKORLANTAS POLRI\nこ   
gambar_filtering_23.jpg                           \r\nB 5859 FLE\nYAMAHA\nMAX\nLIVE IN THE HIGHER STAGE-\nYAMAHA\n02.20   
gambar_filtering_24.jpg                                                  \r\nIt.\n700\nC\nÖRLANT\nj\n5395 FFE\n10-26\n>   
gambar_filtering_25.jpg                                     \r\nB 5199 FSP\n05.29\nPOLRI\nMIX\nLIVE IN THE HIGHER STAGE   
gambar_filtering_26.jpg                                                                   \r\nB 5946 FCO\nVAMAHA\n02-26   
gambar_filtering_27.jpg                                                                           \r\nB 5039 TLX\n10.27   
gambar_filtering_28.jpg                                                                \r\nC\nB 4318 FFB\nHONDA\n[06.26   
gambar_filtering_29.jpg                                                               \r\nB 4542 TPX\nYOSHIMURALL 06.28   
gambar_filtering_3.jpg                                                   \r\nگا\n30\nZZ\n10\nΟ\n"\n28\nKOR\nPOLRI\nے\n♡   
gambar_filtering_30.jpg                                                                        \r\nB 3416 TPL\n09.27\nا   
gambar_filtering_31.jpg                                                                            \r\nB4739 FWD\n08.29   
gambar_filtering_32.jpg                                                                 \r\nB 5100 FLW\n^\nHONDA\n12.28   
gambar_filtering_33.jpg                                                                   \r\nB 4238 KIW\nHONDA\n(06-28   
gambar_filtering_34.jpg                                   